docs/proposals/error-handling-guideline.md

```diff
@@ -0,0 +1,221 @@
+Error: unknown distribution specification flag: v1.0
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type oras/test localhost:7000/command/images:foobar --distribution-spec ???
+Error: unknown distribution specification flag: "v1.0". 
+Available options: v1.1-referrers-api, v1.1-referrers-tag
+```
+
+#### Example 6: when attaching an file, if no file reference is provided
+
+Current behavior and output:
+
+```
+$ oras attach --artifact-type oras/test /tmp/ginkgo2977244222:foobar >>
+Error: no blob is provided
+```
+
+Suggested error message:
+
+```
+$ oras attach --artifact-type oras/test /tmp/ginkgo2977244222:foobar >>
+Error: no blob is provided
+Usage: oras attach [flags] --artifact-type=<type> <name>{:<tag>|@<digest>} <file>[:<type>] [...]
+To attach files to an existing artifact, try "oras attach --artifact-type sbom/example oras/test /tmp/ginkgo2977244222:foobar sample.txt". Run "oras attach -h" for more options and examples
+```
+
+#### Example 7: When pushing files, if the annotation file doesn't match the required format
+
+Current behavior and output:
+
+```
+$ oras push --annotation-file sbom.json ghcr.io/library/alpine:3.9
+Error: failed to load annotations from sbom.json: json: cannot unmarshal string into Go value of type map[string]map[string]string. Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations.
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation-file annotation.json ghcr.io/library/alpine:3.9
+Error: failed to load annotations from annotation.json: annotation file or syntax doesn't match the required format or syntax
+Please refer to the document at https://oras.land/docs/how_to_guides/manifest_annotations.
+```
+
+#### Example 8: When pushing files, if the annotation value doesn't match the required syntax
+
+Current behavior and output:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: missing key in `--annotation` flag: key:value
+```
+
+Suggested error message:
+
+```
+$ oras push --annotation "key:value" ghcr.io/library/alpine:3.9
+Error: annotation value doesn't match the required format.
+Try `oras push --annotation "key=value" ghcr.io/library/alpine:3.9`  
+```
+
+#### Example 9: when failed to pull files from a public registry
+
+```
+$ oras pull docker.io/nginx:latest
+Error: failed to resolve latest: GET "https://registry-1.docker.io/v2/nginx/manifests/latest": response status code 401: unauthorized: authentication required: [map[Action:pull Class: Name:nginx Type:repository]]
+```
+
+Suggested error message:
+
+```
+$ oras pull docker.io/nginx:latest
+Error: failed to resolve the resource from server: pull access denied for ghcr.io/nginx:latest : response status code 401: unauthorized: requested access to the resource is denied
+Namespace is missing, do you mean `oras pull docker.io/library/nginx:latest`? 
+```
+
+## Reference
```

Comment:
One thing that is not covered in this spec is to specify which errors will output error code 0 (automation will continue) and which will output something different from 0 (automation will break). In my experience, ORAS returns error codes that break automation when they shouldn't. The typical example is when manifest index delete fails for OCI 1.0 references.

Reply:
Good point. I will investigate current behavior of ORAS, what is the right approach to return error code 0, and when to break automation. Then add a paragraph for this part

Reply:
I think my ask here is different. What you point me to is what I explained above with 0 and 1 Exist Status. What I am asking is to have a guidance what type of errors will return `EXIT_SUCCESS` and `EXIT_FAILURE`. I am envisioning one specific use case where DELETE of the manifest index fails but the reference is pushed and from customer point of view this is `EXIT_SUCCESS` but we return `EXIT_FAILURE`

Reply:
@toddysm 
All types of errors will be in stderr and should return `EXIT_FAILURE`, the automation should also be terminated.

For that specific case, I remember it was reproduced in `oras attach` when use ORAS with registries that don't support Deletion API (e.g. Docker Hub). It was a bug and has been fixed. The reference will not be attached successfully and `oras attach` will fail in that case.

@shizhMSFT can correct me if I am wrong.

Reply:
Here's the definition: The `oras` CLI returns `EXIT_FAILURE` if and only if the `oras` CLI reports one or more errors.

In the case of returning an "error" with `EXIT_SUCCESS`, that "error" is considered as a warning and shall be corrected as a warning.

---

> I am envisioning one specific use case where DELETE of the manifest index fails but the reference is pushed and from customer point of view this is `EXIT_SUCCESS` but we return `EXIT_FAILURE`

In the above case the deletion failed shall be corrected (if not corrected already) to a warning instead of an error, and thus the `oras` CLI returns `EXIT_SUCCESS`.

Reply:
Thanks @shizhMSFT for the clarification. I added your suggested definition to the General guiding principles section.